src/style/style_layer/custom_style_layer.js

```diff
@@ -0,0 +1,182 @@
+ *         gl.attachShader(this.program, vertexShader);
+ *         gl.attachShader(this.program, fragmentShader);
+ *         gl.linkProgram(this.program);
+ *     }
+ * 
+ *     render(gl, matrix) {
+ *         gl.useProgram(this.program);
+ *         gl.uniformMatrix4fv(gl.getUniformLocation(this.program, "u_matrix"), false, matrix);
+ *         gl.drawArrays(gl.POINTS, 0, 1);
+ *     }
+ * }
+ *
+ * map.on('load', function() {
+ *     map.addLayer(new NullIslandLayer());
+ * });
+ */
+
+/**
+ * Called when the layer has been added to the Map with {@link Map#addLayer}. This
+ * gives the layer a chance to initialize gl resources and register event listeners.
+ *
+ * @function
+ * @memberof CustomLayerInterface
+ * @instance
+ * @name onAdd
+ * @param {Map} map The Map this custom layer was just added to.
+ * @param {WebGLRenderingContext} gl The gl context for the map.
+ */
+
+/**
+ * Called when the layer has been removed from the Map with {@link Map#removeLayer}. This
+ * gives the layer a chance to clean up gl resources and event listeners.
+ *
+ * @function
+ * @memberof CustomLayerInterface
+ * @instance
+ * @name onRemove
+ * @param {Map} map The Map this custom layer was just added to.
+ */
+
+/**
+ * Called during a render frame allowing the layer to draw 2D features into the GL context.
+ *
+ * The layer cannot make any assumptions about the current GL state.
+ * The layer must not change any state related to the depth buffer.
+ *
+ * @function
+ * @memberof CustomLayerInterface
+ * @instance
+ * @name render3D
+ * @param {WebGLRenderingContext} gl The map's gl context.
+ * @param {Array<number>} matrix The map's camera matrix. It projects spherical mercator
+ * coordinates to gl coordinates. The spherical mercator coordinate `[0, 0]` represents the
+ * top left corner of the mercator world and `[1, 1]` represents the bottom right corner. The z coordinate
+ * is conformal. A box with identical x, y, and z lengths in mercator units would be rendered as
+ * a cube.
+ */
+
+/**
+ * Called during a render frame allowing the layer to draw 3D features into the GL context.
+ *
+ * The layer cannot make any assumptions about the current GL state.
+ * The layer must not change any state related to the depth buffer.
+ * Correct feature order is only supported for opaque features.
+ *
+ * @function
+ * @memberof CustomLayerInterface
+ * @instance
+ * @name render
+ * @param {WebGLRenderingContext} gl The map's gl context.
+ * @param {Array<number>} matrix The map's camera matrix. It projects spherical mercator
+ * coordinates to gl coordinates. The spherical mercator coordinate `[0, 0]` represents the
+ * top left corner of the mercator world and `[1, 1]` represents the bottom right corner.
+ */
+export type CustomLayerInterface = {
+    id: string,
+    type: "custom",
+    render: CustomRenderMethod,
+    render3D: CustomRenderMethod,
+    onAdd: (map: Map, gl: WebGLRenderingContext) => void,
+    onRemove(map: Map): void
+}
```

Comment:
This type definition says that `render` and `render3D` are both defined. Should they both be typed `?CustomRenderMethod` instead?